NDTensors/src/dense/dense.jl

```diff
@@ -3,73 +3,145 @@
+  if N > 1
+    println("Only Vector-based datatypes are currently supported.")
+    throw(TypeError)
   end
 
-  # Special case of Array which gets turned into a Vector
-  function Dense{ElT,VecT}(data::Array) where {ElT,VecT<:AbstractArray{ElT}}
-    return new{ElT,VecT}(vec(data))
+  return Dense{ElT,typedDataT}(generic_zeros(typedDataT, dim(inds)))
+end
+
+function Dense{ElT,DataT}(dim::Integer) where {ElT,DataT<:AbstractArray{<:Any,N}} where {N}
+  typedDataT = set_eltype_if_unspecified(DataT, ElT)
+  if N > 1
+    println("Only vector-based datatypes are currently supported.")
+    throw(TypeError)
   end
+  return Dense{ElT,typedDataT}(generic_zeros(typedDataT, dim))
 end
 
-function Dense(data::VecT) where {VecT<:AbstractArray{ElT}} where {ElT}
-  return Dense{ElT,VecT}(data)
+function Dense{ElT,DataT}(
+  ::UndefInitializer, inds::Tuple
+) where {ElT,DataT<:AbstractArray{<:Any,N}} where {N}
+  typedDataT = set_eltype_if_unspecified(DataT, ElT)
+  if N > 1
+    println("Only Vector-based datatypes are currently supported.")
+    throw(TypeError)
+  end
+  return Dense{ElT,typedDataT}(similar(typedDataT, dim(inds)))
 end
 
-function Dense(data::Array{ElT}) where {ElT}
-  return Dense{ElT,Vector{ElT}}(vec(data))
+function Dense{ElR,DataT}(data::AbstractArray) where {ElR,DataT<:AbstractArray}
+  if eltype(data) != ElR
+    data = similartype(typeof(data), ElR)(data)
+  end
+  return Dense{ElR,DataT}(data)
 end
 
-function Dense{ElR}(data::AbstractArray{ElT}) where {ElR,ElT}
-  return ElT == ElR ? Dense(data) : Dense(ElR.(data))
+function Dense{DataT}() where {DataT<:AbstractArray}
+  typedDataT = set_eltype_if_unspecified(DataT, default_eltype())
+  ElT = eltype(typedDataT)
+  return Dense{ElT,typedDataT}()
 end
 
-# Construct from a set of indices
-function Dense{ElT,VecT}(inds) where {ElT,VecT<:AbstractArray{ElT}}
-  return Dense(VecT(dim(inds)))
+function Dense{DataT}(inds::Tuple) where {DataT<:AbstractArray}
+  typedDataT = set_eltype_if_unspecified(DataT)
+  ElT = eltype(typedDataT)
+  return Dense{ElT,typedDataT}(inds)
 end
 
-Dense{ElT}(dim::Integer) where {ElT} = Dense(zeros(ElT, dim))
+function Dense{DataT}(::UndefInitializer, inds::Tuple) where {DataT<:AbstractArray}
+  typedDataT = set_eltype_if_unspecified(DataT)
+  ElT = eltype(typedDataT)
+  return Dense{ElT,typedDataT}(undef, inds)
+end
 
-Dense{ElT}(::UndefInitializer, dim::Integer) where {ElT} = Dense(Vector{ElT}(undef, dim))
+function Dense{DataT}(x::Number, inds::Tuple) where {DataT<:AbstractArray}
+  typedDataT = set_eltype_if_unspecified(DataT, typeof(x))
+  ElT = eltype(typedDataT)
+  return Dense{ElT,typedDataT}(fill!(similar(typedDataT, dim(inds)), x))
+end
 
-Dense(::Type{ElT}, dim::Integer) where {ElT} = Dense{ElT}(dim)
+# This function is ill-defined. It cannot transform a complex type to real...
+function Dense{ElR}(data::AbstractArray{ElT}) where {ElR,ElT}
+  return Dense(similartype(typeof(data), ElR)(data))
+end
```

Comment:
Note that we should generally use `convert` for this since it doesn't copy data if it doesn't need to, and also I would call `Dense{ElR}` inside with the pattern that the constructors should keep getting more specific, i.e.:
```julia
function Dense{ElR}(data::AbstractArray) where {ElR}
  return Dense{ElR}(convert(similartype(typeof(data), ElR), data))
end
```

Reply:
Of course there should then be a version:
```julia
function Dense{ElR}(data::AbstractArray{ElR}) where {ElR}
  # [...]
end
```
which catches when the element type is already correct, otherwise we end up in an infinite loop.

Reply:
Okay, I have updated this. Thanks!